### PR TITLE
46269 backend [ users ] Hierarchical approval workflow

### PR DESCRIPTION
**Pull request by EBirkenfeld**
## 1. Description (Problem)

The system lacked a "Manager of Step N" performer type — the ability to assign a workflow step to the manager of the user who completed a specified previous step. Previously, achieving a similar effect required manually assigning specific users or using workarounds.

**Where it manifested:** in the Template Editor, in the Performers dropdown on the step form — no "Manager: <step name>" option was available.

## 2. Context

This feature was needed to support hierarchical business processes where a task must be routed to a specific performer's supervisor. Example: an employee completes "Prepare Report" → the next step "Approval" is automatically assigned to their manager.

Ticket: **#46269**. This extends the `PerformerType` model following the same pattern as existing types `workflow_starter` and `field`.

## 3. Solution

Added a new `PerformerType.MANAGER` performer type across the full stack:

- **Backend:** new enum value, `source_task_api_name` field on `RawPerformer` / `RawPerformerTemplate`, `_resolve_manager()` method on `Task`, validation in the serializer.
- **Frontend:** new `ETaskPerformerType.Manager` enum, generation of "Manager: <step>" dropdown options, reference cleanup on step deletion.

**Key principle — JIT resolution:** the manager is determined **only when the step starts** (`start_task` → `update_performers`), not in advance. If the employee's manager changes during the workflow — the system uses the current one. Resolution queries `WorkflowEvent(type=TASK_COMPLETE)` to find the actual user who completed the source step, then follows `select_related('user__manager')` in a single optimized query. Using `WorkflowEvent` (rather than `TaskPerformer.date_completed`) ensures reliable resolution regardless of performer configuration — groups, multiple users, or mixed setups all work correctly because the event always records the actual completer.

**Key design decisions:**
- Manager is **not mutually exclusive** — can coexist with user/group performers on the same step
- **Multiple Manager performers** on one step — allowed, each resolves independently
- Template steps are **non-linear** (may activate via conditions), so the dropdown shows all steps, not just "previous"
- `require_completion_by_all` on the source step is **prohibited** by validation (no single completer can be determined)
- Reject logic is implemented via standard conditions + radio + revert_task, no new code needed

### Backend (8 files + 1 migration)

| File | Change |
|------|--------|
| `processes/enums.py` | Added `PerformerType.MANAGER = 'manager'` to choices |
| `processes/models/mixins.py` | Added `source_task_api_name` field to `RawPerformerMixin`; updated validation — `MANAGER` does not require user/group/field |
| `processes/models/workflows/task.py` | Added `source_task_api_name` to `_get_raw_performer()`, `add_raw_performer()`, and `_get_raw_performers_template()`; new `_resolve_manager()` method using `WorkflowEvent(type=TASK_COMPLETE)` with `select_related('user__manager')`; `MANAGER` handling in `update_performers()` |
| `processes/serializers/templates/raw_performer.py` | Validation `additional_validate_raw_performers_type_manager()` (source_id presence, self-reference, missing step, `require_completion_by_all`); `to_representation()` — `label: "Manager: <step>"`; `create()/update()` — mapping `source_id → source_task_api_name` |
| `processes/serializers/templates/task.py` | Passes `template_tasks` context (including `require_completion_by_all`) when manager performers are present |
| `processes/messages/template.py` | 4 new validation messages: `MSG_PT_0071`–`MSG_PT_0074` |
| `migrations/0252_add_manager_performer_type.py` | `AddField source_task_api_name` on `RawPerformerTemplate` and `RawPerformer` |

### Frontend (11 files)

| File | Change |
|------|--------|
| `types/template.ts` | Added `Manager = 'manager'` to `ETaskPerformerType` |
| `getPerformersForDropdown.ts` | Generates manager options from the list of steps (excluding current) |
| `TaskPerformers.tsx` | Passes `tasks` prop; forms `value` with `manager-` prefix |
| `TaskForm.tsx` | Passes `tasks` to `<TaskPerformers>` |
| `TaskItemUsers.tsx` | Renders `<Avatar isEmpty />` for manager type |
| `UsersDropdown.tsx` | Added `EOptionTypes.Manager` |
| `template.ts` | Cleanup logic: removes manager performers when source step is missing |
| `en_US.ts` / `ru_RU.ts` | Localization: `'Manager: {step}'` / `'Менеджер: {step}'` |
| `.storybook/preview.tsx` | Translated comment RU→EN |
| `RichEditor.spec.test.tsx` | Translated describe blocks RU→EN |

### Tests

- **Backend:** `test_resolve_manager.py` — 13 tests: 8 performer configuration cases (`single_user`, `two_users`, `single_group`, `group_and_user`, `two_groups`, `two_groups_two_users`, `user_in_own_group`, `empty_group`) + 5 edge cases (`source_not_completed`, `no_manager`, `empty_api_name`, `nonexistent_api_name`, `completer_user_deleted`); `test_task.py` — 2 tests for `add_raw_performer` (`manager_with_source_api_name`, `user_without_source_api_name`); updated mocks in `test_task.py` (5 fixes: +`source_task_api_name=None`)
- **Frontend:** `getPerformersForDropdown.test.ts` — 4 tests (multiple steps, no tasks, single step, unique value); `template.test.ts` — 3 tests (deleted source step, step exists, empty sourceId); updated `TaskPerformers.test.tsx`

### 5.1 Preconditions

- Account with at least 2 users on the same team, one assigned as the other's manager (Settings → Team → select Manager for a user)
- Template with at least 2 steps
- Dev environment

### 5.2 Positive Scenarios

1. **Creating a template with a Manager performer:**
   - [ ] Open Template Editor → create a template with 2+ steps
   - [ ] On Step 2, open the Performers dropdown
   - [ ] Verify dropdown contains "Manager: <Step 1 name>" options (and other steps, excluding current)
   - [ ] Select "Manager: Step 1" → save template
   - [ ] **Expected:** template saves without errors, performer is displayed in the list

2. **Running a workflow and manager resolution:**
   - [ ] Start a workflow from the created template
   - [ ] The subordinate user completes Step 1
   - [ ] **Expected:** Step 2 is assigned to that subordinate's manager

3. **Display in TaskItem (step card):**
   - [ ] In the template step list, verify that the manager performer displays as an empty avatar (same as Workflow Starter)

4. **Chain of managers ("manager of manager"):**
   - [ ] Create a template with 3 steps: Step 1 → regular user, Step 2 → Manager of Step 1, Step 3 → Manager of Step 2
   - [ ] Run workflow: Employee → their Manager → Manager's Manager (Director)
   - [ ] **Expected:** each level receives the task correctly

5. **Manager + regular performers on the same step:**
   - [ ] Add both Manager: Step 1 and a regular user as performers on the same step
   - [ ] Run workflow
   - [ ] **Expected:** both performers are assigned, step works with all of them

### 5.3 Negative Scenarios and Edge Cases

1. **Self-reference:**
   - [ ] Attempt to add a Manager performer referencing the step's own step
   - [ ] **Expected:** validation error "Manager performer cannot reference its own step"

2. **Non-existent step (deleted):**
   - [ ] Create a template, add Manager: Step 1, then delete Step 1
   - [ ] **Expected:** performer is automatically removed from the list (cleanTemplateReferences)

3. **Step with require_completion_by_all:**
   - [ ] Enable "Required completion by all" on the source step, try to add a Manager performer on another step referencing it
   - [ ] **Expected:** validation error "Disable 'Required completion by all' on step..."

4. **No manager assigned + no other performers:**
   - [ ] Run a workflow where the subordinate (who completed the step) has no manager assigned, and the target step has no other performers
   - [ ] **Expected:** the step is skipped (SKIPPED)

5. **No manager assigned + other performers exist:**
   - [ ] Same as above, but the target step also has a regular user performer
   - [ ] **Expected:** the step is NOT skipped, works with remaining performers

6. **Deactivated manager:**
   - [ ] The user's manager is deactivated (inactive)
   - [ ] **Expected:** treated as "no manager" — fallback to remaining performers or skip

7. **Source step not yet completed:**
   - [ ] In a non-standard scenario where the source step is still active (conditions led to this)
   - [ ] **Expected:** manager not found, fallback behavior

8. **Group performer completes source step:**
   - [ ] Source step has a GROUP performer (not individual user). A member of the group completes the step.
   - [ ] **Expected:** Manager of the group member who actually clicked "Complete" is resolved correctly

9. **Mixed performers on source step (group + user):**
   - [ ] Source step has both a GROUP and individual USER performers. Any of them completes.
   - [ ] **Expected:** Manager is resolved for the actual completer, regardless of performer type

### 5.4 Verification Points

- **UI:** Performers dropdown contains Manager options; selected performer displays correctly after page reload
- **API:** in template response (`GET /templates/{id}`), rawPerformer with `type: "manager"` has `source_id` and `label: "Manager: <step>"` fields
- **DB:** in `processes_rawperformertemplate` / `processes_rawperformer` tables, `source_task_api_name` field is populated

### 5.5 API Checks

- **Saving template (PUT/PATCH `/templates/{id}`):**
  - In DevTools → Network → find the save request
  - In request body (`payload`) → `tasks[N].rawPerformers` → for manager performer: `type: "manager"`, `source_id: "<step api_name>"`

- **Getting template (GET `/templates/{id}`):**
  - In response `tasks[N].rawPerformers` → for manager performer: `type: "manager"`, `source_id: "<api_name>"`, `label: "Manager: <step name>"`

### 5.6 What Was NOT Tested

- Locales partially verified (en/ru labels added, full locale switching was not tested)
- Not tested on mobile devices
- Not tested in production (verified in dev only)
- Not tested under load (many steps, many manager performers)
- Delegation (Substitutes): if the manager is on vacation and delegated — standard substitutes logic is expected to work, but not explicitly verified
- Custom task view for manager (context showing "whose task", approve/reject buttons) — planned for Phase 2
- System variable "who completed the step" — not implemented, planned for Phase 2

## 6. Testing Affected Areas (Dependencies)

| Component | What to verify |
|-----------|---------------|
| `TaskPerformers` | Basic dropdown functionality is intact — User, Group, Workflow Starter, Field selection still works |
| `TaskItemUsers` | Performer rendering in step cards is correct for all types |
| `cleanTemplateReferences()` | Field performers still cleaned correctly when steps are deleted |
| `_get_raw_performer()` | Existing types (user, group, field, workflow_starter) are created without regressions |
| `update_performers()` | Calls `_resolve_manager` — main runtime path where manager is assigned |
| `RawPerformerSerializer` validation | MSG_PT_0074 correctly rejects MANAGER on steps with `require_completion_by_all=True`; valid templates save without errors |

## 7. Refactoring

- Translated comments in `.storybook/preview.tsx` (RU→EN)
- Translated `describe` blocks in `RichEditor.spec.test.tsx` (RU→EN)
- These changes are cosmetic and do not affect logic

## 8. Release Notes

**New Feature: "Manager of Step" Performer Type**

Templates now support a "Manager" performer type. When configuring a workflow step, you can assign it to the manager of the user who completed a specific previous step. This enables hierarchical approval flows where tasks automatically route to the appropriate supervisor.

---

> [!NOTE]
> **Medium Risk**
> Adds a new performer type that dynamically assigns tasks based on workflow completion events and introduces new persisted fields via migration; incorrect resolution or edge cases (reverts, inactive managers) could change performer assignment at runtime.
> 
> **Overview**
> Adds a new `PerformerType.MANAGER` that assigns a task to the **manager of the user who completed a specified source step**.
> 
> Backend adds `source_task_api_name` to raw performer models (and versioning schema), validates manager performers at template-save time (required source step, no self-reference, source must exist, and cannot be `require_completion_by_all`), and resolves the manager during `Task.update_performers()` by looking up the latest `TASK_COMPLETE` workflow event and clearing stale performer links when resolution fails.
> 
> Frontend updates the template editor to offer **“Manager: {step}”** options in the performers dropdown (unique `manager-` values), display manager performers in task cards, localize labels, and extend template cleanup to drop manager performers if their source step is removed. Includes comprehensive backend/frontend tests for resolution, serialization/versioning, and UI option generation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8f10699d9686e3998db9e910a55527fde9f8c1f4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add hierarchical approval workflow with Manager performer type
> - Introduces a new `manager` performer type that dynamically assigns the manager of the user who completed a specified source step as the performer for a subsequent task.
> - Adds `source_task_api_name` field to `RawPerformer` and `RawPerformerTemplate` models (via migration [0252](https://github.com/pneumaticapp/pneumaticworkflow/pull/210/files#diff-3c93fd3eef17af2ab0ca239b85706f32ca79e07e06708b7d5a4c8038b7d31d27)) to link a manager performer to its source step.
> - Backend resolver (`Task._resolve_manager`) looks up the completer of the source task via workflow events and resolves their manager; if no manager is found or the source task is reverted, the TaskPerformer link is cleared.
> - Frontend dropdown in [getPerformersForDropdown.ts](https://github.com/pneumaticapp/pneumaticworkflow/pull/210/files#diff-ab5a3b4dfb674561a7d7bd98145c62441ef7bc880b368faf2ab55aba68994628) now lists manager-of-step options (excluding the current step), and [cleanTemplateReferences](https://github.com/pneumaticapp/pneumaticworkflow/pull/210/files#diff-bba6bc9d61c0008252857e2bf1f5e624bea1f2e72358cebea80327e1bf25a960) removes manager performers whose source step no longer exists.
> - Risk: manager performer resolution is dynamic and depends on workflow event history; reverting a source task clears the manager assignment, which may be unexpected for in-progress workflows.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 8f10699.</sup>
> 
>
>

### CODE REVIEW

**Review by pneumatic-ilyacedrik — DISMISSED** [review]

File: backend/src/processes/models/workflows/task.py [hunk 1]
```diff
@@ -151,13 +158,17 @@ def add_raw_performer(
...
         performer_type: PerformerType = PerformerType.USER,
+        source_task_api_name: Optional[str] = None,
     ):  # -> RawPerformer
 
         """ Creates and returns a raw performer for a task with given data
             Optionally updates the performers after create raw performers """
 
         if (
-            performer_type != PerformerType.WORKFLOW_STARTER
+            performer_type not in (
```
Comment (pneumojoseph) [inline backend/src/processes/models/workflows/task.py]: An exception should also be raised for PerformerType.MANAGER if source_task_api_name is not passed.

File: backend/src/processes/models/workflows/task.py [hunk 2]
```diff
@@ -318,6 +336,38 @@ def update_raw_performers_from_task_template(
...
+            the source step and return their manager.
+            Returns None if source step not completed
+            or completer has no manager. """
+
+        source_api_name = raw_performer.source_task_api_name
+        if not source_api_name:
+            return None
+        from src.processes.models.workflows.event import WorkflowEvent
+        complete_event = (
+            WorkflowEvent.objects
```
Comment (pneumojoseph) [inline backend/src/processes/models/workflows/task.py:356]: If the Task returns and starts again, there will be several events, you need to find the newest event

File: backend/src/processes/models/workflows/task.py [hunk 3]
```diff
@@ -151,13 +158,17 @@ def add_raw_performer(
...
+        source_task_api_name: Optional[str] = None,
     ):  # -> RawPerformer
 
         """ Creates and returns a raw performer for a task with given data
             Optionally updates the performers after create raw performers """
 
         if (
-            performer_type != PerformerType.WORKFLOW_STARTER
+            performer_type not in (
+                PerformerType.WORKFLOW_STARTER,
```
Comment (pneumojoseph) [inline backend/src/processes/models/workflows/task.py]: I'm not sure the process update for PerformerType.MANAGER is working. RawPerformerTemplateSchemaV1 doesn't contain the new source_task_api_name field. Improve your test coverage.

File: backend/src/processes/models/workflows/task.py (hunk 2 above)
Comment (EBirkenfeld) [reply backend/src/processes/models/workflows/task.py:356]: Already handled — .order_by('-created').first() returns the newest event, even if the task went through revert+recomplete and there are multiple events. The test test_resolve_manager__revert_recomplete__returns_latest_manager covers this scenario (user_a completes first → revert → user_b completes again → returns user_b's manager).

**Review by pneumojoseph — APPROVED** [review]
